File: dataforge-plots/plots-fx/src/main/java/hep/dataforge/plots/fx/FXPlotManager.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.plots.fx;

import hep.dataforge.context.BasicPlugin;
import hep.dataforge.context.PluginDef;
import hep.dataforge.exceptions.NameNotFoundException;
import hep.dataforge.fx.FXPlugin;
import hep.dataforge.meta.Laminate;
import hep.dataforge.meta.Meta;
import hep.dataforge.plots.PlotFrame;
import hep.dataforge.plots.PlotManager;
import javafx.application.Platform;
import javafx.scene.Scene;
import javafx.stage.Stage;
import javafx.stage.Window;

import java.util.HashMap;
import java.util.Map;
import java.util.ServiceLoader;
import java.util.stream.StreamSupport;

/**
 * A plot holder using MDI-style JavaFX containers
 *
 * @author Alexander Nozik
 */
@PluginDef(name = "plots", group = "hep.dataforge", dependsOn = {"hep.dataforge:fx"}, info = "Basic plottiong plugin")
public class FXPlotManager extends BasicPlugin implements PlotManager {
    public static final String FX_FRAME_TYPE_KEY = "fxFrame.type";

    private static ServiceLoader<FXPlotFrameFactory> fxPlotFrameFactoryServiceLoader = ServiceLoader.load(FXPlotFrameFactory.class);

    private final Map<String, Map<String, PlotContainer>> stages = new HashMap<>();

    /**
     * Build an FX frame of the given type using spi.
     *
     * @return
     */
    public static FXPlotFrame buildFXPlotFrame(Meta meta) {
        String type = meta.getString(FX_FRAME_TYPE_KEY, "");
        return StreamSupport.stream(fxPlotFrameFactoryServiceLoader.spliterator(), false)
                .filter(it -> type.isEmpty() || it.getName().equals(type))
                .sorted()
                .findFirst()
                .orElseThrow(() -> new NameNotFoundException(type))
                .build(meta);
    }

    private synchronized Map<String, PlotContainer> getStage(String stage) {
        return stages.computeIfAbsent(stage, stageName -> new HashMap<>());
    }

    protected FXPlotFrame buildFrame(Meta meta) {
        return buildFXPlotFrame(new Laminate(meta, meta()));
    }

    protected synchronized PlotContainer showPlot(String name, FXPlotFrame frame) {
        FXPlugin fx = getContext().getFeature(FXPlugin.class);
        PlotContainer container = FXPlotUtils.displayContainer(fx, name, 800, 600);
        container.setPlot(frame);
        return container;
    }

    public synchronized PlotContainer getPlotContainer(String stage, String name) {
        return getStage(stage).computeIfAbsent(name, plotName -> {
            FXPlotFrame frame = buildFrame(meta());
            return showPlot(name, frame);
        });
    }

    @Override
    public PlotFrame getPlotFrame(String stage, String name) throws NameNotFoundException {
        PlotContainer container = getPlotContainer(stage, name);
        Platform.runLater(() -> {
            Scene scene = container.getPane().getScene();
            if (scene != null) {
                Window window = container.getPane().getScene().getWindow();
                if (window != null && !window.isShowing() && window instanceof Stage) {
                    ((Stage) window).show();
                }
            }
        });
        return container.getPlot();
    }

    @Override
    public boolean hasPlotFrame(String stage, String name) {
        return stages.containsKey(stage) && stages.get(stage).containsKey(name);
    }

}

File: dataforge-gui/src/main/kotlin/hep/dataforge/fx/plots/FXPlotManager.kt
package hep.dataforge.fx.plots

import hep.dataforge.context.BasicPlugin
import hep.dataforge.context.Plugin
import hep.dataforge.context.PluginDef
import hep.dataforge.context.PluginFactory
import hep.dataforge.fx.FXPlugin
import hep.dataforge.fx.output.FXDisplay
import hep.dataforge.fx.output.buildDisplay
import hep.dataforge.meta.Meta
import hep.dataforge.plots.PlotFrame
import hep.dataforge.plots.PlotPlugin
import hep.dataforge.plots.jfreechart.JFreeChartFrame

const val DEFAULT_STAGE_NAME = "";
const val DEFAULT_PLOT_NAME = "";


@PluginDef(name = "plots-fx", group = "hep.dataforge", dependsOn = ["hep.dataforge:fx"], info = "Basic JavaFX plottiong plugin")
class FXPlotManager(meta: Meta = Meta.empty()) : BasicPlugin(meta), PlotPlugin {
    private val stages: MutableMap<String, MutableMap<String, PlotContainer>> = HashMap()

    /**
     * Plot frame factory
     */
    private var frameFactory: () -> PlotFrame = {
        try {
            //initializing fx surrogate if fx framework not started
            context.get(FXPlugin::class.java).checkApp()
            JFreeChartFrame()
        } catch (ex: Exception) {
            throw RuntimeException("Plot frame factory not configured", ex);
        }
    };

    private val display: FXDisplay by lazy {
        buildDisplay(context)
    };

    override fun getPlotFrame(stage: String, name: String): PlotFrame {
        return stages
                .computeIfAbsent(stage) { HashMap() }
                .computeIfAbsent(name) {
                    PlotContainer(frameFactory()).apply {
                        display(stage, name, this)
                    }
                }
                .frame
    }

    private fun display(stage: String, name: String, container: PlotContainer) {
        display.display(stage, name) {
            center = container.root;
        }
//            (pane.scene.window as? Stage)?.show()

    }

    override fun hasPlotFrame(stage: String, name: String): Boolean {
        return stages.containsKey(stage) && (stages[stage]?.containsKey(name) ?: false);
    }

    fun display(stage: String = DEFAULT_STAGE_NAME, name: String = DEFAULT_PLOT_NAME, action: PlotFrame.() -> Unit): PlotFrame {
        return getPlotFrame(stage, name).apply(action)
    }

    fun display(stage: String = DEFAULT_STAGE_NAME, name: String = DEFAULT_PLOT_NAME, frame: PlotFrame): PlotFrame {
        return stages
                .computeIfAbsent(stage) { HashMap() }
                .computeIfAbsent(name) {
                    PlotContainer(frame).apply {
                        display(stage, name, this)
                    }
                }.frame
    }

    class Factory : PluginFactory() {
        override val type: Class<out Plugin> = FXPlotManager::class.java

        override fun build(meta: Meta): Plugin {
            return FXPlotManager(meta)
        }
    }

}